File: app/src/main/java/c4gnv/com/thingsregistry/net/ServiceCallback.java
package c4gnv.com.thingsregistry.net;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public abstract class ServiceCallback<T> implements Callback<T> {

    private int statusCode;

    public abstract void onSuccess(T response);

    public abstract void onFail(ServiceError error);

    @Override
    public void onResponse(Call<T> call, Response<T> response) {
        this.statusCode = response.code();

        if (response.isSuccessful()) {
            onSuccess(response.body());
        } else {
            try {
                String errorJson = response.errorBody().string();
                handleConnectionError(call, statusCode, errorJson);
            } catch (Exception e) {
                handleException(call, e);
            }
        }
    }

    private void handleConnectionError(Call<T> call, int responseCode, String errorJson) {
        onFail(ServiceError.createError(call.request(), ServiceError.ErrorType.CONNECTION));
    }

    private void handleException(Call<T> call, Exception e) {
        onFail(ServiceError.createError(call.request(), ServiceError.ErrorType.UNKNOWN));
    }

    @Override
    public void onFailure(Call<T> call, Throwable t) {
        onFail(ServiceError.from(call, t));
    }
}

File: app/src/main/java/c4gnv/com/thingsregistry/net/ServiceCallback.kt
package c4gnv.com.thingsregistry.net

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

abstract class ServiceCallback<T> : Callback<T> {

    private var statusCode: Int = 0

    abstract fun onSuccess(response: T)

    abstract fun onFail(error: ServiceError)

    override fun onResponse(call: Call<T>, response: Response<T>) {
        this.statusCode = response.code()

        if (response.isSuccessful) {
            onSuccess(response.body() as T)
        } else {
            try {
                val errorJson = response.errorBody()!!.string()
                handleConnectionError(call, statusCode, errorJson)
            } catch (e: Exception) {
                handleException(call, e)
            }

        }
    }

    private fun handleConnectionError(call: Call<T>, responseCode: Int, errorJson: String) {
        onFail(ServiceError.createError(call.request(), ServiceError.ErrorType.CONNECTION))
    }

    private fun handleException(call: Call<T>, e: Exception) {
        onFail(ServiceError.createError(call.request(), ServiceError.ErrorType.UNKNOWN))
    }

    override fun onFailure(call: Call<T>, t: Throwable) {
        onFail(ServiceError.from(call, t))
    }
}
